### DISCUSSION TITLE
Javascript DOM Interview Questions

### DISCUSSION THREAD

**Discussion by anburocky3** [anburocky3/Javascript-DOM-in-Tamil#1] [General]
Share your Interview questions about Javascript DOM.

You can search on the **internet**, **get it from recruiters**, and share it here. We will make a solution along with the video explaining it.

Format example: https://github.com/anburocky3/javascript-interview-questions/blob/main/javascript-arrays/questions.md

**Comment by shirokid** [comment]
## Questions from [this site](https://www.thatjsdude.com/interview/dom.html)
### DOM questions for intermediate JavaScript Developers.

**1. Is there any difference between `window` and `document`?
2. Does `document.onload` and `window.onload` fire at the same time?
3. Is attribute `similiar` to `property`?
4. What are the different ways to get an element from DOM?
5. What is the fastest way to select elements by using css selectors?
6. What is event bubble? How does event flows in DOM?
7. If you need to implement `getElementByAttribute`, how would you implement it?
8. How could you capture all clicks in a page?
9. How can you get all the texts in a web page
10. What is `defer` and `async` keyword does in a script tag?
11. What is repaint and when does this happen?
12. How could we verify whether one element is child of another?
13. How would you add a class to an element by query selector?
14. Why can't we use `forEach` or similiar array methods on a `NodeList`?**

**Comment by anburocky3** [comment]
15. What is DOM?
16. What are the levels involved in DOM?
17. What is a Javascript and how JavaScript worked with DOM(Document Object Model)?
18. Difference between HTML DOM vs Javascript DOM
19. Difference between `window` and `document`?
20. Difference between `window.onload` and `document.onload`?
21. What is the difference between an `attribute` and a `property`?
22. What are the methods to get one or more elements from the DOM?
23. What’s the fastest way to query DOM?
24. How do we use forEach and other array methods on a NodeList?
25. How do we implement a function to get elements by attribute?
26. How HTML DOM allows javascript to change the text color and background color of a particular element?
27. How do we implement a function to add a class to an HTML element?
28. How can DOM change the content or text of a P element?
29. What is the difference between JavaScript and jQuery explain with example?
30. What is a DOM EventListener method and how it can show the current date and time?
31. What is the HTML DOM collection, Explain with example? 
32. How can I find the number of elements (length) in the HTML collection?
33. Write the program to change the background color of the body element by DOM event?

**Comment by anburocky3** [comment]
34. How to insert dom elements in a performant way?
35. How to remove elements from DOM tree?
36. How can you find child elements in a DOM tree?
37. How can you find parent & sibling elements in a DOM tree?
38. How do you write inline styles in DOM tree?
39. What is event listeners in Javascript?
40. How to add an event listener in DOM tree?
41. What is Event bubbling & Event capturing?
42. What is event object?
43. How do you prevent default event actions that is happening inside the document.
44. What is event delegation and write an sample program explaining it.
45. What type of events available? List few.
46. How do you handle input events?
47. Using fetch method display your IP address 
48. What is DOM Nodes?
49. async vs defer?
50. What is querySelector & querySelectorAll?

**Comment by anburocky3** [comment]
## Questions finalized:

| S.no | Question                                                                                                    | Reference                                                                                                                                    |
| ---- | ----------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------- |
| 1    | What is DOM?                                                                                                | [📽️ What is HTML DOM?](https://www.youtube.com/watch?v=D04yjfmgiD4&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=2)                          |
| 2    | What are the levels involved in DOM?                                                                        | [📽️ What is HTML DOM?](https://www.youtube.com/watch?v=D04yjfmgiD4&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=2)                          |
| 3    | What is a Javascript and how JavaScript worked with DOM(Document Object Model)?                             | [📽️ What is HTML DOM?](https://www.youtube.com/watch?v=D04yjfmgiD4&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=2), [What is Javascript](#) |
| 4    | Difference between `HTML DOM` vs `Javascript DOM`?                                                          | [📽️ What is HTML DOM?](https://www.youtube.com/watch?v=D04yjfmgiD4&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=2)                          |
| 5    | Difference between `window` and `document`? DOM?                                                            | [📽️ Window vs Document object](https://youtu.be/PwH3qYb1g1o?list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9)                                         |
| 6    | Does `document.onload` and `window.onload` fire at the same time?                                           | [📽️ DOM Loaded Events](https://www.youtube.com/watch?v=DCvpLOaFkZU&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=8)                          |
| 7    | Is `attribute` similiar to `property`?                                                                      | [📽️ Element Properties vs HTML Attributes](https://www.youtube.com/watch?v=_CJsFQCh4-8&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=29)     |
| 8    | What is `defer` and `async` keyword does in a script tag?                                                   | [📽️ Javascript Scripts: async vs defer](https://www.youtube.com/watch?v=9WHrYavh-hM&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=9)         |
| 9    | What are the different ways to get an `element` from `DOM`?                                                 | [📽️ How to Select DOM Elements](https://www.youtube.com/watch?v=j1UAeKE9rd4&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=11)                |
| 10   | What is the fastest way to `select elements` by using `css selectors`?                                      | [📽️ getElementById() - JS DOM Selector](https://www.youtube.com/watch?v=khi85Y4ylCw&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=12)        |
| 11   | What is DOM Nodes?                                                                                          | [📽️ Handling Input Elements](https://www.youtube.com/watch?v=SlLTuWYa7UM&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=10&t=3s)              |
| 12   | What is `NodeList` in DOM?                                                                                  | [📽️ Understanding DOM Nodes](https://www.youtube.com/watch?v=SlLTuWYa7UM&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=10)                   |
| 13   | Why can't we use forEach or similiar array methods on a NodeList?                                           | [📽️ Understanding DOM Nodes](https://www.youtube.com/watch?v=SlLTuWYa7UM&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=10)                   |
| 14   | How do we implement a function to get elements by attribute?                            | [📽️ Setting & Getting HTML Attributes](https://www.youtube.com/watch?v=q_gQldvWjk8&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=30)         |
| 15   | How can you get all the texts in a web page?                                                                | [📽️ Looping over DOM Elements ](https://www.youtube.com/watch?v=lemV0hgeFFs&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=25)                |
| 16   | How `HTML DOM` allows javascript to change the `text color` and `background color` of a particular element? | [📽️ Setting & Getting Inline Styles](https://www.youtube.com/watch?v=Rf_82qUg58o&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=31&t=40s)     |
| 17   | What is querySelector & querySelectorAll?                                                                   | [📽️ QuerySelector & QueryselectorAll](https://www.youtube.com/watch?v=H1MSkS_mAPk&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=15)          |
| 18   | How would you add a class to an element by query selector?                                                  | [📽️ Setting & Getting Class](https://www.youtube.com/watch?v=M5Bl5dMCn34&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=32)                   |
| 19   | How can DOM change the content or text of a P element?                                                      | [📽️ Understanding DOM Nodes](https://www.youtube.com/watch?v=SlLTuWYa7UM&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=10)                   |
| 20   | What is repaint and when does this happen?                                                                  | [📽️ Javascript DocumentFragment](https://www.youtube.com/watch?v=EddrT4FLzd0&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=19)               |
| 21   | How could we verify whether one element is child of another?                                                | [📽️ Finding Parent DOM Elements ](https://www.youtube.com/watch?v=jt9ikcNGdKU&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=27)              |
| 22   | How to find parent element of nearest ancestor?                                                             | [📽️ Understanding DOM Nodes](https://www.youtube.com/watch?v=SlLTuWYa7UM&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=10)                   |
| 23   | What is the difference between JavaScript and jQuery explain with example?                                  | [📽️ What is HTML DOM?](https://www.youtube.com/watch?v=D04yjfmgiD4&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=2)                          |
| 24   | What is the HTML DOM collection, Explain with example?                                                      | [📽️ Looping over DOM Elements ](https://www.youtube.com/watch?v=lemV0hgeFFs&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=25)                |
| 25   | How can I find the number of elements (length) in the HTML collection?                                      | [📽️ Looping over DOM Elements ](https://www.youtube.com/watch?v=lemV0hgeFFs&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=25)                |
| 26   | Write the program to change the background color of the body element by DOM event?                          | [📽️ Setting & Getting Inline Styles](https://www.youtube.com/watch?v=Rf_82qUg58o&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=31&t=40s)     |
| 27   | How to insert dom elements in a performant way?                                                             | [📽️ Javascript DocumentFragment](https://www.youtube.com/watch?v=EddrT4FLzd0&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=19)               |
| 28   | How to remove elements from DOM tree?                                                                       | [📽️ Removing DOM Elements](https://www.youtube.com/watch?v=AHQqt19ogVE&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=24)                     |
| 29   | How can you find child elements in a DOM tree?                                                              | [📽️ Finding Child DOM Elements](https://www.youtube.com/watch?v=bziRqACn0F4&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=26)                |
| 30   | How can you find parent & sibling elements in a DOM tree?                                                   | [📽️ Finding Parent & Sibling DOM Elements](https://www.youtube.com/watch?v=jt9ikcNGdKU&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=27)     |
| 31   | What is a DOM EventListener method and how it can show the current date and time?                           | [📽️ Adding Event Listeners ](https://www.youtube.com/watch?v=Mtkyq0jXRL8&list=PL73Obo20O_7jhOOPDASWk0PVcRxGEyrm9&index=34&t=958s)            |
| 32   | What is Event bubbling & Event capturing?                                                                   | [📽️ Event Bubbling & Capturing](#)                                                                                                           |
| 33   | What is event object?                                                                                       | [📽️ Event Object](#)                                                                                                                         |
| 34   | How do you prevent default event actions that is happening inside the document.                             | [📽️ Preventing Default Browser Actions](#)                                                                                                   |
| 35   | What is event delegation and write an sample program explaining it.                                         | [📽️ Event Delegation](#)                                                                                                                     |
| 36   | What type of events available? List few.                                                                    | [📽️ Keyboard Events](#)                                                                                                                      |
| 37   | How do you handle input events?                                                                             | [📽️ Handling Input Elements](#)                                                                                                              |
| 38   | Using fetch method display your IP address                                                                  | [📽️ Handling Submit Event & FormData](#)                                                                                                     |
| 39   | Collect user's hobbies from checkbox and log it in console.                                                 | [📽️ Handling Checkbox Elements](#)                                                                                                           |
| 40   | Write a simple form with Javascript Validation                                                              | [📽️ Javascript Form Validation](#)                                                                                                           |
